pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/ModelParser.groovy

```diff
@@ -139,16 +142,38 @@ class ModelParser implements Parser {
     public @CheckForNull ModelASTPipelineDef parse(ModuleNode src, boolean secondaryRun = false) {
         // first, quickly ascertain if this module should be parsed at all
         def pst = src.statementBlock.statements.find {
-            MethodCallExpression m = matchMethodCall(it)
-            return m != null && matchMethodName(m) == ModelStepLoader.STEP_NAME
+            return isDeclarativePipelineStep(it)
         }
 
-        if (pst==null) {
-            // Check if there's a 'pipeline' step somewhere nested within the other statements and error out if that's the case.
-            src.statementBlock.statements.each { checkForNestedPipelineStep(it) }
-            return null; // no 'pipeline', so this doesn't apply
+        if (pst != null) {
+            return parsePipelineStep(pst, secondaryRun)
+        } else {
+            // Look for the pipeline step inside methods named call.
+            MethodNode callMethod = src.methods.find { it.name == "call" }
+            if (callMethod != null) {
+                PipelineStepFinder finder = new PipelineStepFinder()
+                callMethod.code.visit(finder)
+                List<Statement> pipelineSteps = finder.pipelineSteps
+
+                if (!pipelineSteps.isEmpty()) {
+                    List<ModelASTPipelineDef> pipelineDefs = pipelineSteps.collect { p ->
+                        return parsePipelineStep(p, secondaryRun)
+                    }
+                    // Even if there are multiple pipeline blocks, just return the first one - this return value is only
+                    // used in a few places: tests, where there will only ever be one, and linting/converting, which also
+                    // are guaranteed to only have one, since we don't intend to support linting of pipelines defined in
+                    // shared libraries.
```

Comment:
So why waste time parsing them all?

Reply:
Because I can't tell which one is going to be executed. Basically, the return value of `parse` is irrelevant except for those very few places, none of which are applicable to shared libraries in the first place, so we can keep just returning the first one we parse. But we do need to parse all of them in case of the `if (whatever) { pipeline { ... } } else { pipeline { ... } }` case.